README.md
[![Build Status](https://travis-ci.org/masiuchi/p5-Plack-Middleware-WOVN.svg?branch=master)](https://travis-ci.org/masiuchi/p5-Plack-Middleware-WOVN) [![Coverage Status](https://img.shields.io/coveralls/masiuchi/p5-Plack-Middleware-WOVN/master.svg?style=flat)](https://coveralls.io/r/masiuchi/p5-Plack-Middleware-WOVN?branch=master)
# NAME

Plack::Middleware::WOVN - Translates PSGI application by using WOVN.io.

# SYNOPSYS

    use Plack::Builder;

    builder {
      'WOVN',
        settings => {
          user_token => 'token',
          secret_key => 'sectet',
        };
      $app;
    };

# DESCRIPTION

This is a Plack Middleware component for translating PSGI application by using WOVN.io.
Before using this middleware, you must sign up and configure WOVN.io.

# SETTINGS

## user\_token

User token of your WOVN.io account. This value is required.

## secret\_key

This value will be used in the future. But this value is required.

## url\_pattern

URL rewriting pattern of translated page.

- path (default)

        original: http://example.com/

        translated: http://example.com/ja/

- subdomain

        original: http://example.com/

        translated: http://ja.exmple.com/

- query

        original: http://example.com/

        translated: http://example.com/?wovn=ja

## url\_pattern\_reg

This value is coufigured by url\_pattern. You don't have to configure this value.

## query

Filters query parameters when rewriting URL. Default values is \[\]. (Do not filter query)

## api\_url

URL of WOVN.io API. Default value is "https://api.wovn.io/v0/values".

## default\_lang

Default language of web application. Default value is "en".

## supported\_langs

This value will be used in the future. Default value is \["en"\].

## test\_mode

When "on" or "1" is set to "test\_mode", this middleware translates only the page whose url is "test\_url".
Default value is "0".

## test\_url

Default value is not set.

# LICENSE

Copyright (C) 2016 by Masahiro Iuchi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Masahiro Iuchi


Changes
Revision history for Perl extension Plack-Middleware-WOVN

{{$NEXT}}

0.09 2017-04-07T00:00:00Z

    - Fix wrong license.

0.08 2016-05-25T03:01:33Z

    - Fix failed test. #20

0.07 2016-05-24T15:13:13Z

    - Change HTML parser. #19
    - Fix empty parent node error. #18
    - Fix broken regexp of query mode. #15
    - Support Bulgarian. #14

0.06 2016-03-12T09:26:34Z

    - Add LWP::Protocol::https. #12
    - Do not decode JSON when parsing. #12

0.05 2016-03-07T13:49:01Z

    - Remove Try::Tiny from Plack::Middleware::WOVN. #8
    - Fix 307 redirect bug again. #7

0.04 2016-03-06T16:18:30Z

    - Fix 307 redirect does not work. #6

0.03 2016-02-26T07:45:59Z

    - original version



lib/Plack/Middleware/WOVN/Lang.pm
package Plack::Middleware::WOVN::Lang;
use strict;
use warnings;
use utf8;

#http://msdn.microsoft.com/en-us/library/hh456380.aspx
our $LANG = {
    'ar' => { name => 'ﺎﻠﻋﺮﺒﻳﺓ', code => 'ar', en => 'Arabic' },
    'zh-CHS' =>
        { name => '简体中文', code => 'zh-CHS', en => 'Simp Chinese' },
    'zh-CHT' =>
        { name => '繁體中文', code => 'zh-CHT', en => 'Trad Chinese' },
    'da' => { name => 'Dansk',            code => 'da', en => 'Danish' },
    'nl' => { name => 'Nederlands',       code => 'nl', en => 'Dutch' },
    'en' => { name => 'English',          code => 'en', en => 'English' },
    'fi' => { name => 'Suomi',            code => 'fi', en => 'Finnish' },
    'fr' => { name => 'Français',        code => 'fr', en => 'French' },
    'de' => { name => 'Deutsch',          code => 'de', en => 'German' },
    'el' => { name => 'Ελληνικά', code => 'el', en => 'Greek' },
    'he' => { name => 'עברית',       code => 'he', en => 'Hebrew' },
    'id' => { name => 'Bahasa Indonesia', code => 'id', en => 'Indonesian' },
    'it' => { name => 'Italiano',         code => 'it', en => 'Italian' },
    'ja' => { name => '日本語',        code => 'ja', en => 'Japanese' },
    'ko' => { name => '한국어',        code => 'ko', en => 'Korean' },
    'ms' => { name => 'Bahasa Melayu',    code => 'ms', en => 'Malay' },
    'no' => { name => 'Norsk',            code => 'no', en => 'Norwegian' },
    'pl' => { name => 'Polski',           code => 'pl', en => 'Polish' },
    'pt' => { name => 'Português',       code => 'pt', en => 'Portuguese' },
    'ru' => { name => 'Русский',   code => 'ru', en => 'Russian' },
    'es' => { name => 'Español',         code => 'es', en => 'Spanish' },
    'sv' => { name => 'Svensk',           code => 'sv', en => 'Swedish' },
    'th' => { name => 'ภาษาไทย', code => 'th', en => 'Thai' },
    'hi' => { name => 'हिन्दी',    code => 'hi', en => 'Hindi' },
    'tr' => { name => 'Türkçe', code => 'tr', en => 'Turkish' },
    'uk' =>
        { name => 'Українська', code => 'uk', en => 'Ukrainian' },
    'vi' => { name => 'Tiếng Việt', code => 'vi', en => 'Vietnamese' },
};

sub get_code {
    my ( $class, $lang_name ) = @_;
    return undef unless defined $lang_name;
    return $lang_name if $LANG->{$lang_name};

    $lang_name = lc $lang_name;

    for my $key ( keys %$LANG ) {
        if (   $lang_name eq lc $LANG->{$key}{name}
            || $lang_name eq lc $LANG->{$key}{en}
            || $lang_name eq lc $LANG->{$key}{code} )
        {
            return $LANG->{$key}{code};
        }
    }

    return undef;
}

sub get_lang {
    my ( $class, $lang ) = @_;
    my $lang_code = $class->get_code($lang);
    $lang_code ? $LANG->{$lang_code} : undef;
}

1;

__END__

=head1 NAME

Plack::Midleware::WOVN::Lang - Language code list that can be used by WOVN API.

=head1 SEE ALSO

L<Plack::Middleware::WOVN>

=cut



lib/Plack/Middleware/WOVN/Store.pm
package Plack::Middleware::WOVN::Store;
use strict;
use warnings;
use utf8;
use parent 'Class::Accessor::Fast';

use JSON;
use LWP::UserAgent;

use Plack::Middleware::WOVN::Lang;

__PACKAGE__->mk_accessors(qw/ config_loaded /);

my $default_settings = {
    user_token      => '',
    secret_key      => '',
    url_pattern     => 'path',
    url_pattern_reg => "/(?<lang>[^/.?]+)",
    query           => [],
    api_url         => 'https://api.wovn.io/v0/values',
    default_lang    => 'en',
    supported_langs => ['en'],
    test_mode       => 0,
    test_url        => '',
};

sub new {
    my $class = shift;
    my $self  = $class->SUPER::new();

    my $args     = shift             || {};
    my $settings = $args->{settings} || {};

    $self->{settings} = +{ %$default_settings, %$settings };

    $self;
}

sub is_valid_settings {
    my $self            = shift;
    my $settings        = $self->{settings};
    my $user_token      = $settings->{user_token} || '';
    my $secret_key      = $settings->{secret_key};
    my $url_pattern     = $settings->{url_pattern};
    my $query           = $settings->{query};
    my $api_url         = $settings->{api_url};
    my $default_lang    = $settings->{default_lang};
    my $supported_langs = $settings->{supported_langs};

    my $valid = 1;
    my @errors;

    if ( length $user_token < 5 || length $user_token > 6 ) {
        $valid = 0;
        push @errors, "User token $user_token is not valid.";
    }
    if ( !defined $secret_key || $secret_key eq '' ) {
        $valid = 0;
        $secret_key = '' unless defined $secret_key;
        push @errors, "Secret key $secret_key is not valid.";
    }
    if ( !defined $url_pattern || $url_pattern eq '' ) {
        $valid = 0;
        $url_pattern = '' unless defined $url_pattern;
        push @errors, "Url pattern $url_pattern is not valid.";
    }
    if ( !$query || ref $query ne 'ARRAY' ) {
        $valid = 0;
        $query = '' unless defined $query;
        push @errors, "query config $query is not valid.";
    }
    if ( !defined $api_url || $api_url eq '' ) {
        $valid = 0;
        push @errors, 'API url is not configured.';
    }
    if ( !defined $default_lang || $default_lang eq '' ) {
        $valid = 0;
        $default_lang = '' unless defined $default_lang;
        push @errors, "Default lang $default_lang is not valid.";
    }
    if (   !$supported_langs
        || ref $supported_langs ne 'ARRAY'
        || @$supported_langs < 1 )
    {
        $valid = 0;
        push @errors, 'Supported langs configuration is not valid.';
    }

    if (@errors) {

        # output error logs?
    }

    $valid;
}

sub settings {
    my $self     = shift;
    my $settings = $self->{settings};

    if ( !$self->config_loaded ) {
        $settings->{default_lang} = Plack::Middleware::WOVN::Lang->get_code(
            $settings->{default_lang} );
        if ( !exists $settings->{supported_langs} ) {
            $settings->{supported_langs} = [ $settings->{default_lang} ];
        }
        if ( $settings->{url_pattern} eq 'path' ) {
            $settings->{url_pattern_reg} = "/(?<lang>[^/.?]+)";
        }
        elsif ( $settings->{url_pattern} eq 'query' ) {
            $settings->{url_pattern_reg}
                = "((\\?.*&)|\\?)wovn=(?<lang>[^&]+)(&|\$)";
        }
        elsif ( $settings->{url_pattern} eq 'subdomain' ) {
            $settings->{url_pattern_reg} = "^(?<lang>[^.]+)\.";
        }
        if ( !$settings->{test_mode} || $settings->{test_mode} ne 'on' ) {
            $settings->{test_mode} = 0;
        }
        else {
            $settings->{test_mode} = 1;
        }

        $self->config_loaded(1);
    }

    $settings;
}

sub get_values {
    my ( $self, $url ) = @_;
    $url =~ s/\/+$//;

    my $api_url    = $self->{settings}{api_url};
    my $user_token = $self->{settings}{user_token};

    my $ua  = LWP::UserAgent->new;
    my $res = $ua->get("$api_url?token=$user_token&url=$url");

    my $values = {};
    if ( $res->is_success ) {
        $values = eval { JSON::from_json( $res->content ) } || {};
    }

    $values;
}

1;

__END__

=head1 NAME

Plack::Middleware::WOVN::Store - Retrieves data from WOVN API server.

=head1 SEE ALSO

L<Plack::Middleware::WOVN>

=cut



lib/Plack/Middleware/WOVN.pm
package Plack::Middleware::WOVN;
use strict;
use warnings;
use utf8;
use parent 'Plack::Middleware';

our $VERSION = '0.08';

require bytes;

use HTML::HTML5::Parser;
use HTML::HTML5::Writer;
use Mojo::URL;
use Plack::Util;
use Plack::Util::Accessor qw( settings );
use URI::Escape;
use XML::LibXML;

use Plack::Middleware::WOVN::Headers;
use Plack::Middleware::WOVN::Lang;
use Plack::Middleware::WOVN::Store;

our $STORE;

sub prepare_app {
    my $self = shift;
    $STORE = Plack::Middleware::WOVN::Store->new(
        { settings => $self->settings } );
}

sub call {
    my ( $self, $env ) = @_;

    unless ( $STORE->is_valid_settings ) {
        return $self->app->($env);
    }

    my $headers
        = Plack::Middleware::WOVN::Headers->new( $env, $STORE->settings );
    if (   $STORE->settings->{test_mode}
        && $STORE->settings->{test_url} ne $headers->url )
    {
        return $self->app->($env);
    }

    if ( $headers->path_lang eq $STORE->settings->{default_lang} ) {
        my $redirect_headers
            = $headers->redirect( $STORE->settings->{default_lang} );
        return [ 307, [%$redirect_headers], [''] ];
    }
    my $lang = $headers->lang_code;

    my $res = $self->app->( $headers->request_out );
    Plack::Util::response_cb(
        $res,
        sub {
            my $res = shift;

            sub {
                my $body_chunk  = shift or return;
                my $status      = $res->[0];
                my $res_headers = $res->[1];

                if ((   Plack::Util::header_get( $res_headers,
                            'Content-Type' )
                        || ''
                    ) =~ /html/
                    )
                {
                    my $values = $STORE->get_values( $headers->redis_url );
                    my $url    = {
                        protocol => $headers->protocol,
                        host     => $headers->host,
                        pathname => $headers->pathname,
                    };
                    $body_chunk
                        = switch_lang( $body_chunk, $values, $url, $lang,
                        $headers )
                        unless $status =~ /^1|302/;
                }

                Plack::Util::header_set( $res_headers, 'Content-Length',
                    bytes::length $body_chunk );

                $body_chunk;
            };
        }
    );
}

sub add_lang_code {
    my ( $href, $pattern, $lang, $headers ) = @_;
    return $href if $href =~ /^(#.*)?$/;

    my $new_href = $href;
    my $lc_lang  = lc $lang;

    if ( $href && lc($href) =~ /^(https?:)?\/\// ) {
        my $uri = eval { Mojo::URL->new($href) } or return $new_href;

        if ( lc $uri->host eq lc $headers->host ) {
            if ( $pattern eq 'subdomain' ) {
                my $sub_d = $href =~ /\/\/([^\.]*)\./ ? $1 : '';
                my $sub_code
                    = Plack::Middleware::WOVN::Lang->get_code($sub_d);
                if ( $sub_code && lc $sub_code eq $lc_lang ) {
                    $new_href =~ s/$lang/$lc_lang/i;
                }
                else {
                    $new_href =~ s/(\/\/)([^\.]*)/$1$lc_lang\.$2/;
                }
            }
            elsif ( $pattern eq 'query' ) {
                if ( $href =~ /\?/ ) {
                    $new_href = "$href&wovn=$lang";
                }
                else {
                    $new_href = "$href?wovn=$lang";
                }
            }
            else {
                $new_href =~ s/([^\.]*\.[^\/]*)(\/|$)/$1$lang/;
            }
        }
    }
    elsif ($href) {
        if ( $pattern eq 'subdomain' ) {
            my $lang_url
                = $headers->protocol . '://'
                . $lc_lang . '.'
                . $headers->host;
            my $current_dir = $headers->pathname;
            $current_dir =~ s/[^\/]*\.[^\.]{2,6}$//;
            if ( $href =~ /^\.\..*$/ ) {
                $new_href =~ s/^(\.\.\/)+//;
                $new_href = $lang_url . '/' . $new_href;
            }
            elsif ( $href =~ /^\..*$/ ) {
                $new_href =~ s/^(\.\/)+//;
                $new_href = $lang_url . $current_dir . '/' . $new_href;
            }
            elsif ( $href =~ /^\/.*$/ ) {
                $new_href = $lang_url . $href;
            }
            else {
                $new_href = $lang_url . $current_dir . '/' . $href;
            }
        }
        elsif ( $pattern eq 'query' ) {
            if ( $href =~ /\?/ ) {
                $new_href = "$href&wovn=$lang";
            }
            else {
                $new_href = "$href?wovn=$lang";
            }
        }
        else {
            if ( $href =~ /^\// ) {
                $new_href = '/' . $lang . $href;
            }
            else {
                my $current_dir = $headers->pathname;
                $current_dir =~ s/[^\/]*\.[^\.]{2,6}$//;
                $new_href = '/' . $lang . $current_dir . $href;
            }
        }
    }

    $new_href;
}

sub check_wovn_ignore {
    my $node = shift;
    if ( !$node->isa('XML::LibXML::Text') ) {
        if ( defined $node->getAttribute('wovn-ignore') ) {
            $node->setAttribute( 'wovn-ignore', '' )
                if $node->getAttribute('wovn-ignore') eq 'wovn-ignore';
            return 1;
        }
        elsif ( $node->nodeName eq 'html' ) {
            return 0;
        }
    }
    if ( !$node->getParentNode ) {
        return 0;
    }
    check_wovn_ignore( $node->getParentNode );
}

sub switch_lang {
    my ( $body, $values, $url, $lang, $headers ) = @_;
    $lang ||= $STORE->settings->{'default_lang'};
    $lang = Plack::Middleware::WOVN::Lang->get_code($lang);
    my $text_index     = $values->{text_vals}      || {};
    my $src_index      = $values->{img_vals}       || {};
    my $img_src_prefix = $values->{img_src_prefix} || '';
    my $string_index   = {};

    my $tree = HTML::HTML5::Parser->load_html( string => $body );
    $tree->setEncoding('UTF-8');

    my $writer = HTML::HTML5::Writer->new(
        quote_attributes => 1,
        voids            => 1,
        start_tags       => 1,
        end_tags         => 1
    );

    if ( $tree->documentElement->hasAttribute('wovn-ignore') ) {
        $body =~ s/href="([^"]*)"/"href=\"".uri_unescape($1)."\""/eg;
        return $body;
    }

    if ( $lang ne $STORE->settings->{default_lang} ) {
        for my $node ( $tree->findnodes("//*[local-name()='a']") ) {
            next if check_wovn_ignore($node);
            my $href = $node->getAttribute('href');
            my $new_href
                = add_lang_code( $href, $STORE->settings->{url_pattern},
                $lang, $headers );
            $node->setAttribute( 'href', $new_href );
        }
    }

    for my $node ( $tree->findnodes('//text()') ) {
        next if check_wovn_ignore($node);
        my $node_text = $node->getValue;
        $node_text =~ s/^\s+|\s+$//g;
        if (   $text_index->{$node_text}
            && $text_index->{$node_text}{$lang}
            && @{ $text_index->{$node_text}{$lang} } )
        {
            my $data    = $text_index->{$node_text}{$lang}[0]{data};
            my $content = $node->getValue;
            $content =~ s/^(\s*)[\S\s]*(\s*)$/$1$data$2/g;
            $node->setData($content);
        }
    }

    for my $node ( $tree->findnodes("//*[local-name()='meta']") ) {
        next if check_wovn_ignore($node);
        next
            if ( $node->getAttribute('name')
            || $node->getAttribute('property')
            || '' )
            !~ /^(description|title|og:title|og:description|twitter:title|twitter:description)$/;

        my $node_content = $node->getAttribute('content');
        $node_content =~ s/^\s+\|\s+$//g;
        if (   $text_index->{$node_content}
            && $text_index->{$node_content}{$lang}
            && @{ $text_index->{$node_content}{$lang} } )
        {
            my $data    = $text_index->{$node_content}{$lang}[0]{data};
            my $content = $node->getAttribute('content');
            $content =~ s/^(\s*)[\S\s]*(\s*)$/$1$data$2/g;
            $node->setAttribute( 'content', $content );
        }
    }

    for my $node ( $tree->findnodes("//*[local-name()='img']") ) {
        next if check_wovn_ignore($node);
        if ( lc( $writer->element($node) ) =~ /src=['"]([^'"]*)['"]/ ) {
            my $src = $1;
            if ( $src !~ /:\/\// ) {
                if ( $src =~ /^\// ) {
                    $src = $url->{protocol} . '://' . $url->{host} . $src;
                }
                else {
                    $src
                        = $url->{protocol} . '://'
                        . $url->{host}
                        . $url->{path}
                        . $src;
                }
            }

            if (   $src_index->{$src}
                && $src_index->{$src}{$lang}
                && @{ $src_index->{$src}{$lang} } )
            {
                $node->setAttribute( 'src',
                    $img_src_prefix . $src_index->{$src}{$lang}[0]{data} );
            }
        }
        if ( my $alt = $node->getAttribute('alt') ) {
            $alt =~ s/^\s+|\s+$//g;
            if (   $text_index->{$alt}
                && $text_index->{$alt}{$lang}
                && @{ $text_index->{$alt}{$lang} } )
            {
                my $data = $text_index->{$alt}{$lang}[0]{data};
                $alt =~ s/^(\s*)[\S\s]*(\s*)$/$1$data$2/g;
                $node->setAttribute( 'alt', $alt );
            }
        }
    }

    for my $node ( $tree->findnodes("//*[local-name()='script']") ) {
        if (   $node->getAttribute('src')
            && $node->getAttribute('src')
            =~ /\/\/j.(dev-)?wovn.io(:3000)?\// )
        {
            $node->getParentNode->removeChild($node);
        }
    }

    my ($parent_node) = $tree->getElementsByTagName('head');
    ($parent_node) = $tree->getElementByTagName('body') unless $parent_node;
    $parent_node = $tree->doucmentElement unless $parent_node;

    {
        my $insert_node = XML::LibXML::Element->new('script');
        $insert_node->setAttribute( 'src',   '//j.wovn.io/1' );
        $insert_node->setAttribute( 'async', 'true' );
        my $data_wovnio
            = 'key='
            . $STORE->settings->{user_token}
            . '&backend=true&currentLang='
            . $lang
            . '&defaultLang='
            . $STORE->settings->{default_lang}
            . '&urlPattern='
            . $STORE->settings->{url_pattern}
            . '&version='
            . $VERSION;
        $insert_node->setAttribute( 'data-wovnio', $data_wovnio );
        $insert_node->appendText(' ');
        $parent_node->insertBefore( $insert_node, $parent_node->firstChild );
    }

    for my $l ( get_langs($values) ) {
        my $insert_node = XML::LibXML::Element->new('link');
        $insert_node->setAttribute( 'rel',      'alternate' );
        $insert_node->setAttribute( 'hreflang', $l );
        $insert_node->setAttribute( 'href', $headers->redirect_location($l) );
        $parent_node->appendChild($insert_node);
    }

    my $html = $tree->documentElement;
    $html->setAttribute( 'lang', $lang ) if $html;

    my $new_body = $writer->document($tree);
    $new_body =~ s/href="([^"]*)"/'href="'.uri_unescape($1).'"'/eg;

    $new_body;
}

sub get_langs {
    my $values = shift;
    my %langs;
    my %merged
        = ( %{ $values->{text_vals} || {} }, %{ $values->{img_vals} || {} } );
    for my $index ( values %merged ) {
        for my $key ( keys %{ $index || {} } ) {
            $langs{$key} = 1;
        }
    }
    keys %langs;
}

1;

__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::WOVN - Translates PSGI application by using WOVN.io.

=head1 SYNOPSYS

  use Plack::Builder;

  builder {
    'WOVN',
      settings => {
        user_token => 'token',
        secret_key => 'sectet',
      };
    $app;
  };

=head1 DESCRIPTION

This is a Plack Middleware component for translating PSGI application by using WOVN.io.
Before using this middleware, you must sign up and configure WOVN.io.

This is a port of wovnjava (https://github.com/wovnio/wovnjava).

=head1 SETTINGS

=head2 user_token

User token of your WOVN.io account. This value is required.

=head2 secret_key

This value will be used in the future. But this value is required.

=head2 url_pattern

URL rewriting pattern of translated page.

=over 4

=item * path (default)

  original: http://example.com/

  translated: http://example.com/ja/

=item * subdomain

  original: http://example.com/

  translated: http://ja.exmple.com/

=item * query

  original: http://example.com/

  translated: http://example.com/?wovn=ja

=back

=head2 url_pattern_reg

This value is coufigured by url_pattern. You don't have to configure this value.

=head2 query

Filters query parameters when rewriting URL. Default values is []. (Do not filter query)

=head2 api_url

URL of WOVN.io API. Default value is "https://api.wovn.io/v0/values".

=head2 default_lang

Default language of web application. Default value is "en".

=head2 supported_langs

This value will be used in the future. Default value is ["en"].

=head2 test_mode

When "on" or "1" is set to "test_mode", this middleware translates only the page whose url is "test_url".
Default value is "0".

=head2 test_url

Default value is not set.

=head1 LICENSE

MIT License

Copyright (c) 2016 Minimal Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

=head1 AUTHOR

Masahiro Iuchi

=cut
